brian2/codegen/generators/base.py

```diff
@@ -215,10 +216,27 @@ def has_repeated_indices(self, statements):
                          if index not in ('_idx', '0'))
         return not all_unique
 
+    _translate_cache = {}
     def translate(self, code, dtype):
         '''
         Translates an abstract code block into the target language.
         '''
+        cache_key = _hashable([code, dtype, self.variables,
+                               self.variable_indices,
+                               self.override_conditional_write,
+                               self.allows_scalar_write,
+                               self.iterate_all,
+                               self.template_name,
+                               self.name,
+                               self.owner.name,
+                               {var.owner.name
+                                for var in self.variables.itervalues()
+                                if getattr(var, 'owner', None) is not None},
+                               prefs,
+                               self.__class__])
```

Comment:
Alternatively, we consider this sort of change as part of a future larger refactor of code generation, but it makes me feel very wary.